data-loader/core/src/main/java/com/scalar/db/dataloader/core/dataimport/task/ImportTask.java

```diff
@@ -0,0 +1,431 @@
+            .build();
+      }
+    }
+
+    Optional<Result> optionalScalarDBResult;
+
+    try {
+      optionalScalarDBResult =
+          getDataRecord(
+              namespace, table, optionalPartitionKey.get(), optionalClusteringKey.orElse(null));
+    } catch (ScalarDBDaoException e) {
+      return ImportTargetResult.builder()
+          .namespace(namespace)
+          .tableName(table)
+          .status(ImportTargetResultStatus.RETRIEVAL_FAILED)
+          .errors(Collections.singletonList(e.getMessage()))
+          .build();
+    }
+    ImportTaskAction importAction =
+        optionalScalarDBResult.isPresent() ? ImportTaskAction.UPDATE : ImportTaskAction.INSERT;
+
+    if (importAction == ImportTaskAction.INSERT
+        && shouldRevalidateMissingColumns(importOptions, checkForMissingColumns)) {
+      ImportSourceRecordValidationResult validationResultForMissingColumns =
+          new ImportSourceRecordValidationResult();
+      ImportSourceRecordValidator.checkMissingColumns(
+          mutableSourceRecord, columnNames, validationResultForMissingColumns, tableMetadata);
+      if (!validationResultForMissingColumns.isValid()) {
+        return ImportTargetResult.builder()
+            .namespace(namespace)
+            .tableName(table)
+            .status(ImportTargetResultStatus.MISSING_COLUMNS)
+            .errors(
+                Collections.singletonList(
+                    CoreError.DATA_LOADER_UPSERT_INSERT_MISSING_COLUMNS.buildMessage()))
+            .build();
+      }
+    }
+
+    if (shouldFailForExistingData(importAction, importOptions)) {
+      return ImportTargetResult.builder()
+          .namespace(namespace)
+          .tableName(table)
+          .importedRecord(mutableSourceRecord)
+          .importAction(importAction)
+          .status(ImportTargetResultStatus.DATA_ALREADY_EXISTS)
+          .errors(
+              Collections.singletonList(CoreError.DATA_LOADER_DATA_ALREADY_EXISTS.buildMessage()))
+          .build();
+    }
+
+    if (shouldFailForMissingData(importAction, importOptions)) {
+      return ImportTargetResult.builder()
+          .namespace(namespace)
+          .tableName(table)
+          .importedRecord(mutableSourceRecord)
+          .importAction(importAction)
+          .status(ImportTargetResultStatus.DATA_NOT_FOUND)
+          .errors(Collections.singletonList(CoreError.DATA_LOADER_DATA_NOT_FOUND.buildMessage()))
+          .build();
+    }
+
+    List<Column<?>> columns;
+
+    try {
+      columns =
+          ColumnUtils.getColumnsFromResult(
+              optionalScalarDBResult.orElse(null),
+              mutableSourceRecord,
+              importOptions.isIgnoreNullValues(),
+              tableMetadata);
+    } catch (Base64Exception | ColumnParsingException e) {
+      return ImportTargetResult.builder()
+          .namespace(namespace)
+          .tableName(table)
+          .status(ImportTargetResultStatus.VALIDATION_FAILED)
```

Comment:
Probably same as above? Could you take a look at those error messages added by this PR from the view point of users who need to investigate failures?